whproxy/proxy.go

```diff
@@ -0,0 +1,214 @@
+package whproxy
+
+import (
+	"bufio"
+	"io"
+	"net/http"
+	"regexp"
+	"sync"
+
+	"github.com/gorilla/websocket"
+	"github.com/taskcluster/webhooktunnel/util"
+	"github.com/taskcluster/webhooktunnel/wsmux"
+)
+
+var (
+	registerRe = regexp.MustCompile("^/register/(\\w+)/?$")
+	serveRe    = regexp.MustCompile("^/(\\w+)/?(.*)$")
+)
+
+// Config for Proxy. Accepts a websocket.Upgrader and a Logger.
+// Default value for Upgrade ReadBufferSize and WriteBufferSize is 1024 bytes.
+// Default Logger is NilLogger.
+type Config struct {
+	Upgrader websocket.Upgrader
+	Logger   util.Logger
+}
+
+// Proxy is used to send http and ws requests to workers.
+// New proxy can be created by using whproxy.New()
+type Proxy struct {
+	m               sync.RWMutex
+	pool            map[string]*wsmux.Session
+	upgrader        websocket.Upgrader
+	logger          util.Logger
+	handler         http.Handler
+	onSessionRemove func(string)
+}
+
+func (p *Proxy) validateRequest(r *http.Request) error {
+	return nil
+}
+
+// New returns a pointer to a new proxy instance
+func New(conf Config) *Proxy {
+	p := &Proxy{
+		pool:     make(map[string]*wsmux.Session),
+		upgrader: conf.Upgrader,
+		logger:   conf.Logger,
+	}
+
+	if p.logger == nil {
+		p.logger = &util.NilLogger{}
+	}
+
+	p.handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// register will be matched first
+		if registerRe.MatchString(r.URL.Path) { // matches "/register/(\w+)/?$"
+			id := registerRe.FindStringSubmatch(r.URL.Path)[1]
+			p.register(w, r, id)
+		} else if serveRe.MatchString(r.URL.Path) { // matches "/{id}/{path}"
+			matches := serveRe.FindStringSubmatch(r.URL.Path)
+			id, path := matches[1], matches[2]
+			p.serveRequest(w, r, id, path)
+		} else { // if not register request or worker request, not found
+			http.NotFound(w, r)
+		}
+	})
+
+	return p
+}
+
+func (p *Proxy) SetSessionRemoveHandler(h func(string)) {
+	p.m.Lock()
+	defer p.m.Unlock()
+	p.onSessionRemove = h
+}
+
+// GetHandler returns the router assosciated with the proxy
+func (p *Proxy) GetHandler() http.Handler {
+	return p.handler
+}
+
+// getWorkerSession returns true if a session with the given id is present
+func (p *Proxy) getWorkerSession(id string) (*wsmux.Session, bool) {
+	p.m.RLock()
+	defer p.m.RUnlock()
+	s, ok := p.pool[id]
+	return s, ok
+}
+
+// addWorker adds a new worker to the pool
+func (p *Proxy) addWorker(id string, conn *websocket.Conn, config wsmux.Config) error {
```

Comment:
This was implemented before register. I just felt the code reads nicer, but adding this to register should not be a problem.